### PR TITLE
PUBDEV-7473 allow include categorical column name in beta constraints

### PR DESCRIPTION
**Pull request by koniecsveta**
https://h2oai.atlassian.net/browse/PUBDEV-7473

### CODE REVIEW

**Comment by wendycwong** [comment]
Can you add a Python unit test?

File: h2o-algos/src/test/java/hex/glm/GLMTest.java [hunk 1]
```diff
@@ -2288,4 +2288,76 @@ private void isFeatureUsedInPredictHelper(boolean ignoreConstCols, boolean multi
       Scope.exit();
     }
   }
+
```
Comment (wendycwong) [inline h2o-algos/src/test/java/hex/glm/GLMTest.java:2291]: Can you add a python unit test?

**Review by wendycwong — CHANGES_REQUESTED** [review]

File: h2o-algos/src/main/java/hex/DataInfo.java [hunk 2]
```diff
@@ -857,7 +858,47 @@ public final int nextNumericIdx(int currentColIdx) {
...
+        prefix = prefix.substring(0, prefix.length() - 1);
+      }
+      originalNames[j] = prefix;
+      i = end + 1;
+      j++;
+    }
+    return originalNames;
+  }
   
+  public static String findLongestCommonPrefix(String inputArray[]) {
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/DataInfo.java:881]: move to arrayutils?

File: h2o-algos/src/test/java/hex/glm/GLMBetaConstraintsEncoderTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            
+        } finally {
+            Scope.exit();
+        }
+    }
+
+    void testFramesEqual(Frame frame1, Frame frame2) {
+        BufferedString tmpStr = new BufferedString();
+        for (int j = 0; j < frame1.anyVec().length(); j++) {
+            Assert.assertTrue(frame1.vec(0).atStr(tmpStr, j).equals(frame2.vec(0).atStr(tmpStr, j)));
```
Comment (honzasterba) [inline h2o-algos/src/test/java/hex/glm/GLMBetaConstraintsEncoderTest.java:79]: use static imports

File: h2o-algos/src/main/java/hex/glm/GLM.java [hunk 4]
```diff
@@ -3022,7 +3020,23 @@ public double applyBounds(double d, int i) {
...
+      FrameUtils.BetaConstraintsEncoder constraintsEncoder = new FrameUtils.BetaConstraintsEncoder(_dinfo.coefNames(), _dinfo.coefOriginalNames());
+      Frame transformedFrame =  constraintsEncoder.doAll(getBetaConstraintsTypesForEncoder(beta_constraints), beta_constraints).outputFrame();
+      transformedFrame.setNames(beta_constraints._names);
+      return transformedFrame;
+    }
+    
+    private byte[] getBetaConstraintsTypesForEncoder(Frame beta_constraints) {
+      byte[] types = beta_constraints.types();
+      int id = Arrays.asList(beta_constraints.names()).indexOf("names");
+      types[id] = 2;
```
Comment (honzasterba) [inline h2o-algos/src/main/java/hex/glm/GLM.java:3033]: what is 2? why not use constant from Vec?

File: h2o-algos/src/test/java/hex/glm/GLMTest.java [hunk 5]
```diff
@@ -2288,4 +2288,76 @@ private void isFeatureUsedInPredictHelper(boolean ignoreConstCols, boolean multi
...
+          if (betaConstraints.vec("names").atStr(tmpStr, j).toString().equals(names[i])) {
+            if (Math.abs(beta[i] - betaConstraints.vec("lower_bounds").at(j)) < 1e-4 || Math.abs(beta[i] - betaConstraints.vec("upper_bounds").at(j)) < 1e-4) {
+              continue outer;
+            }
+          }
+        }
+        assertEquals(0, grad[i], 1e-2);
+      }
+    } finally {
+      fr.delete();
```
Comment (honzasterba) [inline h2o-algos/src/test/java/hex/glm/GLMTest.java:2358]: why not use Scope?

File: h2o-algos/src/test/java/hex/glm/GLMTest.java (hunk 1 above)
Comment (koniecsveta) [reply h2o-algos/src/test/java/hex/glm/GLMTest.java:2291]: 0275e3c

File: h2o-algos/src/test/java/hex/glm/GLMTest.java (hunk 5 above)
Comment (koniecsveta) [reply h2o-algos/src/test/java/hex/glm/GLMTest.java]: d5106d4

File: h2o-algos/src/main/java/hex/glm/GLM.java (hunk 4 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/glm/GLM.java]: d5106d4

File: h2o-algos/src/test/java/hex/glm/GLMBetaConstraintsEncoderTest.java (hunk 3 above)
Comment (koniecsveta) [reply h2o-algos/src/test/java/hex/glm/GLMBetaConstraintsEncoderTest.java]: d5106d4

File: h2o-algos/src/main/java/hex/DataInfo.java (hunk 2 above)
Comment (koniecsveta) [reply h2o-algos/src/main/java/hex/DataInfo.java]: d5106d4

**Review by honzasterba — APPROVED** [review]

**Review by wendycwong — APPROVED** [review]
Thank you Zuzana.  You give me an R test for bonus!

**Pull request merged**